server/src/server.ts

```diff
@@ -717,6 +724,117 @@ export default class BashServer {
     logger.debug('onWorkspaceSymbol')
     return this.analyzer.findDeclarationsWithFuzzySearch(params.query)
   }
+
+  private onPrepareRename(params: LSP.PrepareRenameParams): LSP.Range | null {
+    const symbol = this.analyzer.symbolAtPointFromTextPosition(params)
+    this.logRequest({ request: 'onPrepareRename', params, word: symbol?.word })
+
+    if (
+      !symbol ||
+      (symbol.type === 'variable' &&
+        (symbol.word === '_' || !/^[a-z_][\w]*$/i.test(symbol.word)))
+    ) {
+      return null
+    }
+
+    if (Builtins.isBuiltin(symbol.word)) {
+      this.throwResponseError('You cannot rename a built-in command.')
+    }
+
+    if (this.executables.isExecutableOnPATH(symbol.word)) {
+      this.throwResponseError('You cannot rename an executable.')
+    }
+
+    if (ReservedWords.isReservedWord(symbol.word)) {
+      this.throwResponseError('You cannot rename a reserved word.')
+    }
```

Comment:
You mean if someone wanted to do something like this?

```bash
echo="Hello"

echo() {
  printf '%s' "$1"
}

echo $echo
```

I tend to veer away from naming functions and variables using builtin and executable names in my scripts because it might cause confusion, so I thought it would be good to put some sort of guardrail for that. Then again, I guess language servers should be more general and less opinionated about these things, and if someone has a good use case for it, then we should give them the tools to work the way they want.

Renaming builtins and executables already work without these checks using the same rules as any other function or variable. If no definition is found, rename all instances. If a definition is found, rename all instances within the declaration's scope. See example below:

[echo-to-print.webm](https://github.com/bash-lsp/bash-language-server/assets/65783406/edfd8f56-fdb8-4ba7-94d2-7c2361a7c2f5)

I'll remove the checks to allow for this behavior.